arbor/morph/sample_tree.cpp

```diff
@@ -0,0 +1,117 @@
+#include <cmath>
+#include <vector>
+
+#include <arbor/math.hpp>
+
+#include <arbor/morph/error.hpp>
+#include <arbor/morph/sample_tree.hpp>
+
+#include "algorithms.hpp"
+#include "io/sepval.hpp"
+#include "util/span.hpp"
+
+namespace arb {
+
+sample_tree::sample_tree(std::vector<msample> samples, std::vector<size_t> parents) {
+    if (samples.size()!=parents.size()) {
+        throw std::runtime_error(
+            "The same number of samples and parent indices used to create a sample morphology");
+    }
+    reserve(samples.size());
+    for (auto i: util::make_span(samples.size())) {
+        append(parents[i], samples[i]);
+    }
+}
+
+void sample_tree::reserve(size_t n) {
+    samples_.reserve(n);
+    parents_.reserve(n);
+    props_.reserve(n);
+    branch_ids_.reserve(n);
+}
+
+size_t sample_tree::append(size_t p, const msample& s) {
+    if (p>size()) {
```

Comment:
Having the root of the sample tree have parent index 0 as opposed to e.g. an `npos` value is inconsistent with the convention in `morphology` with branch parents. Can we change this?

Also, the current test looks like it admits allowing the addition of a point which is its own parent, which in general should not be allowed.

If there is a concern that starting off with an `append(sample_tree::npos, root_sample)` is too verbose, we could have another append method that does not take a parent argument, but uses the last added point as the parent, or `npos` if there are none.

Reply:
Done.